op-supervisor/supervisor/backend/backend.go

```diff
@@ -441,9 +439,105 @@ func (su *SupervisorBackend) checkAccess(acc types.Access, execAt types.Executin
 	if err != nil {
 		return eth.BlockID{}, err
 	}
+
 	return bl.ID(), nil
 }
 
+type rpcCheckResult struct {
+	block eth.BlockID
+	err   error
+}
+
+func (su *SupervisorBackend) verifyAccessWithRPCInParallel(ctx context.Context, executingDescriptor types.ExecutingDescriptor, acc types.Access, primaryAccessCheck func(types.Access) (eth.BlockID, error)) (eth.BlockID, error) {
+	rpcResultChan := make(chan rpcCheckResult, 1)
+	go func() {
+		msgBlockRPC, err := su.checkAccessWithRPC(ctx, acc)
+		rpcResultChan <- rpcCheckResult{block: msgBlockRPC, err: err}
+	}()
+
+	msgBlockDB, err := primaryAccessCheck(acc)
+	if err != nil {
+		su.logger.Debug("Access-list inclusion check failed", "err", err)
+		return eth.BlockID{}, err
+	}
+
+	rpcResult := <-rpcResultChan
+	if rpcResult.err != nil {
+		su.logger.Error("RPC access check failed", "err", rpcResult.err)
+		return eth.BlockID{}, rpcResult.err
+	}
+	if msgBlockDB != rpcResult.block {
+		su.logger.Error("RPC access check failed, DB access check result did not match rpc access check result", "db_block", msgBlockDB, "rpc_block", rpcResult.block)
+		return eth.BlockID{}, fmt.Errorf("RPC access check failed, DB access check result did not match rpc access check result")
+	}
+	return rpcResult.block, nil
+}
+
+func (su *SupervisorBackend) checkAccessWithRPC(ctx context.Context, acc types.Access) (eth.BlockID, error) {
+	rpcAccessVerifyTimeout := 5 * time.Second
+
+	src, ok := su.syncSources.Get(acc.ChainID)
+	if !ok {
+		return eth.BlockID{}, fmt.Errorf("%w: %v", types.ErrUnknownChain, acc.ChainID)
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, rpcAccessVerifyTimeout)
+	defer cancel()
+
+	blockRef, err := src.BlockRefByNumber(timeoutCtx, acc.BlockNumber)
+	if err != nil {
+		return eth.BlockID{}, fmt.Errorf("failed to get block ref for verifiying access with RPC: %w", err)
+	}
+
+	receipts, err := src.FetchReceipts(timeoutCtx, blockRef.Hash)
```

Comment:
This seems to be prohibitively expensive for this critical path since we are fetching every receipt for single block, although it is logically correct. Maybe we need a custom rpc at our L2 EL client which takes block number and log index and returns needed info.

Reply:
See target issue https://github.com/ethereum-optimism/optimism/issues/15128; we have consensus that implementing this using the existing APIs could still be worth it despite its expensive nature. In order to find out, I would propose we run an interop NAT fuzz/stress test again a kurtosis deploy of this setup and at what tx rate this breaks vs the legacy behavior.

To your point, it's also worth asking whether this could be optimized with a new EL API endpoint in op-geth, op-reth, and the spec.